### PR TITLE
Build: webpack all the html/css

### PR DESCRIPTION
**Pull request by MaikoTan**
~~Sorry I still didn't found how to deal with CSSs, for now, they are directly copied from sources to `dist/` directory.~~

~~And a new problem:
how to setup github pages to host from `dist/`~~

Solved by importing [peaceiris/actions-gh-pages](https://github.com/peaceiris/actions-gh-pages)

### CODE REVIEW

**Comment by quisquous** [comment]
Re: CSS.  I think the CSS also needs to be imported, as per https://webpack.js.org/loaders/css-loader/#getting-started.  You add the `style-loader` and `css-loader` and then import the css file in the js file and it all gets bundled.

**Comment by MaikoTan** [comment]
> Re: CSS.  I think the CSS also needs to be imported, as per https://webpack.js.org/loaders/css-loader/#getting-started.  You add the `style-loader` and `css-loader` and then import the css file in the js file and it all gets bundled.

Yep, it would be fine when the project is an SPA.
But for example, oopsy module has two html/css but only one js entry, it is hardly to do a conditionally importing.

**Comment by quisquous** [comment]
> > Re: CSS.  I think the CSS also needs to be imported, as per https://webpack.js.org/loaders/css-loader/#getting-started.  You add the `style-loader` and `css-loader` and then import the css file in the js file and it all gets bundled.
> 
> Yep, it would be fine when the project is an SPA.
> But for example, oopsy module has two html/css but only one js entry, it is hardly to do a conditionally importing.

I think the options here are dynamic import (e.g. https://github.com/quisquous/cactbot/blob/795ffc45c0c322c21e3068929910fb22208293e2/util/gen_coverage_report.js#L67) or adding a top level js entry for each of the two html/css files that imports the shared js and the specific css.

**Comment by MaikoTan** [comment]
I tried `style-loader` but it doesn't work, I think it is the same problem as [this SO question](https://stackoverflow.com/questions/61516920/style-loader-does-not-inject-css-code-in-head-section-of-index-html).
So I would try [mini-css-extract-plugin](https://github.com/webpack-contrib/mini-css-extract-plugin) instead.
And there is some strange issue on my PC, it always failed on `npm install`, I'll solve that first, and then work on this PR later.

**Comment by MaikoTan** [comment]
@quisquous You might need to regenerate `package-lock.json` after this PR merged, I do a clean re-install so that npm works finally, so mess the whole lock file entirely.

**Comment by quisquous** [comment]
I'll try to take a look at this tomorrow or more likely Wednesday when I have a little bit more time to try it out directly.  Sorry for the delay.

**Comment by quisquous** [comment]
I haven't forgotten about this.  It's just a lot to digest and I've been a bit overwhelmed this last week.

**Review by quisquous — COMMENTED** [review]
Overall looks good to me, and I don't have any particular comments about all the code and config changes.  That all seems good to me.  Sorry about the conflicts.

I did some initial testing just of using the dev folder:

* config: doesn't work anymore from inside of ACT for developers (me) linking to the cactbot plugin in their build folder, as `cactbot/ui/config/config.html` is not usable anymore.  Is it possible to allow this to continue to work in some fashion?
* raidboss: the `lippe` skin option is partially broken as `ui/raidboss/skins/lippe/lippe.css` does not exist in the dist folder to be loaded dynamically by user_config.  Timelines are updated (as that's already in an existing css) but alerts are not changed.
* radar: ui/radar/arrow.png is missing
* eureka: ui/eureka/*.png is missing

I wonder if maybe we need to support `ui/${overlay}/resources/` and copy each of those folders as well into dist? That might be a general solution.  Similarly, the `ui/${overlay}/skins/` folder, even if it only exists for raidboss right now.

The output of `util/publish.sh` looks reasonable to me, although it does look like the source maps (or at least large `.map` files) are included there.

**Comment by quisquous** [comment]
Another thought I had is re: typescript.  Users often copy code from javascript to make triggers.  We probably will need to host the typescript converted javascript files somewhere in gh-pages so that users can write those triggers.  Maybe we need to ship that in the release as well? Although that sort of defeats some of the purpose of this patch.  I don't know.  Something to think about in the future for typescript.

File: ui/dps/rdmty/dps.js [hunk 1]
```diff
@@ -44,44 +47,44 @@ var ____Class0=React.Component;for(var ____Class0____Key in ____Class0){if(____C
         return (
             this.props.perSecond === '---' ? null :
             React.createElement("li", {
-                className: 'row ' + this.props.job.toLowerCase() + (this.props.isSelf ? ' self' : ''), 
-                onClick: this.props.onClick}, 
+                className: 'row ' + this.props.job.toLowerCase() + (this.props.isSelf ? ' self' : ''),
```
[the remark this answers is not in the document]
Comment (MaikoTan) [reply ui/dps/rdmty/dps.js]: Sorry this was changed again by my auto formatting tool, I have already reverted once...

File: ui/oopsyraidsy/oopsy_summary.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import './oopsy_common.js';
```
[the remark this answers is not in the document]
Comment (MaikoTan) [reply ui/oopsyraidsy/oopsy_summary.js:1]: Ah, ye mean same as the css files? Okay~

**Comment by MaikoTan** [comment]
Re: pngs are missing

I think this needs a clean up to change the long long image url in code to something like
```javascript
import Image from 'path/to/the/image';
img.src = Image;
```
So webpack would know the image should be bundled in.

I would just copy them from source now, and would do some test soon, to make sure all the module work properly.

**Comment by MaikoTan** [comment]
> * config: doesn't work anymore from inside of ACT for developers (me) linking to the cactbot plugin in their build folder, as `cactbot/ui/config/config.html` is not usable anymore.  Is it possible to allow this to continue to work in some fashion?

What about like I said in #2387 , add an option in `%APPDATA%\Advanced Combat Tracker\Config\RainbowMage.OverlayPlugin.config.json`, something like `"ConfigPageUrl": "cactbot/dist/ui/config/config.html"`?

**Comment by MaikoTan** [comment]
> * raidboss: the `lippe` skin option is partially broken as `ui/raidboss/skins/lippe/lippe.css` does not exist in the dist folder to be loaded dynamically by user_config.  Timelines are updated (as that's already in an existing css) but alerts are not changed.

I didn't found how does the `lippe.css` work, there is only an option in `raidboss_config.js` but no any mechanic to set a css file?
By the way, webpack's `style-loader` support lazy style:
```javascript
import styles from './skins/lippe/lippe.css';

if (xxx)
  styles.use(); // enable the style

if (xxx)
  styles.unuse(); // remove the style
```
So we can refactor the style loading mechanic like this in the future?

**Comment by quisquous** [comment]
> > * config: doesn't work anymore from inside of ACT for developers (me) linking to the cactbot plugin in their build folder, as `cactbot/ui/config/config.html` is not usable anymore.  Is it possible to allow this to continue to work in some fashion?
> 
> What about like I said in #2387 , add an option in `%APPDATA%\Advanced Combat Tracker\Config\RainbowMage.OverlayPlugin.config.json`, something like `"ConfigPageUrl": "path/to/cactbot/dist/ui/config/config.html"`?

I think my preference would be for this to work without an option, if that's possible.

**Comment by quisquous** [comment]
> > * raidboss: the `lippe` skin option is partially broken as `ui/raidboss/skins/lippe/lippe.css` does not exist in the dist folder to be loaded dynamically by user_config.  Timelines are updated (as that's already in an existing css) but alerts are not changed.
> 
> I didn't found how does the `lippe.css` work, there is only an option in `raidboss_config.js` but no any mechanic to set a css file?

Sorry for the confusion.  It's https://github.com/quisquous/cactbot/blob/main/resources/user_config.js#L284-L296

There's an option with hardcoded names, and then if you pick that name it loads particular css files.  I think the skin css files still need to be loaded at that particular place and time in the code, but it could use a different mechanism with explicit imports if that works.

**Comment by MaikoTan** [comment]
> > > * config: doesn't work anymore from inside of ACT for developers (me) linking to the cactbot plugin in their build folder, as `cactbot/ui/config/config.html` is not usable anymore.  Is it possible to allow this to continue to work in some fashion?
> > 
> > What about like I said in #2387 , add an option in `%APPDATA%\Advanced Combat Tracker\Config\RainbowMage.OverlayPlugin.config.json`, something like `"ConfigPageUrl": "path/to/cactbot/dist/ui/config/config.html"`?
> 
> I think my preference would be for this to work without an option, if that's possible.

Better than hard coding, right?

And by the way, I think it is better to seperate out C# and JS code into two repositories, then webpack JS code to a diretory in C# repo, so keep the current relative path.

**Comment by quisquous** [comment]
Just so that I know you aren't waiting on me to review this, there's still a number of things from [this comment](https://github.com/quisquous/cactbot/pull/2420#pullrequestreview-589936102) to address before this can land, right?

**Comment by MaikoTan** [comment]
> Just so that I know you aren't waiting on me to review this, there's still a number of things from [this comment](https://github.com/quisquous/cactbot/pull/2420#pullrequestreview-589936102) to address before this can land, right?

Yes, there were just too many conflicts so I merged the main branch first.

**Comment by quisquous** [comment]
I think #2973 covers what this PR was doing, so I will close this.  I think the CSS is still left as a followup to that patch.  Thanks for all the work and contributions here.  Even if this particular patch land, this helped a lot towards that one.

**Comment by panicstevenson** [comment]
To echo quisquous' thoughts, I just wanted to add a thank you here as well! I feel bad about stealing your work, but I know you've said you've been busy so I didn't want to leave this as a burden for you. Really though, thank you for the hard work here!!

**Comment by MaikoTan** [comment]
That's Okay. (*￣3￣)╭
I want to thank you to make it landed, as I opened this PR but didn't continue to work on it.

**Pull request closed without merging**